libcasr/src/sarif.rs

```diff
@@ -0,0 +1,404 @@
+
+        let results = self.json.as_object_mut().unwrap()["runs"]
+            .as_array_mut()
+            .unwrap()[0]
+            .as_object_mut()
+            .unwrap()["results"]
+            .as_array_mut()
+            .unwrap();
+        let mut result = Map::new();
+        result.insert("ruleId".to_string(), Value::String(rule_id));
+        result.insert("level".to_string(), Value::String("error".to_string()));
+        let mut message = Map::new();
+        let text = if !report.stdin.is_empty() {
+            format!(
+                "{}: {} < {}",
+                report.execution_class.short_description, report.proc_cmdline, report.stdin
+            )
+        } else {
+            format!(
+                "{}: {}",
+                report.execution_class.short_description, report.proc_cmdline
+            )
+        };
+        message.insert("text".to_string(), Value::String(text));
+        result.insert("message".to_string(), Value::Object(message));
+        let mut locations: Vec<Value> = Vec::new();
+        let mut location = Map::new();
+        let mut physical_loc = Map::new();
+        let mut artifact_loc = Map::new();
+        let mut region = Map::new();
+        let parts: Vec<_> = report.crashline.split(':').map(|s| s.to_string()).collect();
+        if parts.len() != 2 && parts.len() != 3 {
+            return Err(Error::Casr(format!(
+                "Unable to parse crashline: {}",
+                report.crashline
+            )));
+        }
+
+        let source_path = Path::new(&parts[0]);
+        let norm_source_path =
+            if let Ok(norm_source_path) = source_path.strip_prefix(source_root.as_ref()) {
+                norm_source_path
+            } else {
+                source_path
+            };
+
+        artifact_loc.insert(
+            "uri".to_string(),
+            Value::String(norm_source_path.display().to_string()),
+        );
+        artifact_loc.insert(
+            "uriBaseId".to_string(),
+            Value::String("%SRCROOT%".to_string()),
+        );
+        physical_loc.insert("artifactLocation".to_string(), Value::Object(artifact_loc));
+        region.insert(
+            "startLine".to_string(),
+            Value::Number(parts[1].parse::<u32>().unwrap().into()),
+        );
+        if parts.len() == 3 {
+            region.insert(
+                "startColumn".to_string(),
+                Value::Number(parts[2].parse::<u32>().unwrap().into()),
+            );
+        }
+        physical_loc.insert("region".to_string(), Value::Object(region));
+        location.insert("physicalLocation".to_string(), Value::Object(physical_loc));
+        locations.push(Value::Object(location));
+
+        result.insert("locations".to_string(), Value::Array(locations));
+
+        let mut stacks: Vec<Value> = Vec::new();
+        let mut stack = Map::new();
+        let mut frames: Vec<Value> = Vec::new();
+        init_ignored_frames!("cpp", "rust", "python", "go", "java");
```

Comment:
hmm, do we really need a filtered stack trace here? or should we display the full stack trace?

Reply:
I suppose, yes. We don't have source code for filtered functions.